compiler/rustc_codegen_llvm/src/gotoc/logging.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+use crate::btree_map;
+use rustc_serialize::json::*;
+use std::env;
+use std::fs::OpenOptions;
+use std::io::{self, prelude::*};
+
+pub use crate::{rmc_log, rmc_warn};
+
+struct LogLine {
+    log_type: LogType,
+    message: String,
+}
+
+impl ToJson for LogLine {
+    fn to_json(&self) -> Json {
+        let output = btree_map![
+            ("log_type".to_string(), self.log_type.to_json()),
+            ("message".to_string(), self.message.to_json())
+        ];
+
+        Json::Object(output)
+    }
+}
+
+enum LogType {
+    Log,
+    Warning(WarningType),
+}
+
+impl ToJson for LogType {
+    fn to_json(&self) -> Json {
+        match self {
+            LogType::Log => {
+                let output = btree_map![("type".to_string(), "LOG".to_json())];
+                Json::Object(output)
+            }
+            LogType::Warning(warning_type) => {
+                let output = btree_map![
+                    ("type".to_string(), "WARNING".to_json()),
+                    ("warning_kind".to_string(), warning_type.to_json())
+                ];
+                Json::Object(output)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum WarningType {
+    Unsupported,
+    Other,
+}
+
+impl ToJson for WarningType {
+    fn to_json(&self) -> Json {
+        format!("{:?}", self).to_json()
+    }
+}
+
+fn write_to_log_file(log_line: LogLine) -> io::Result<()> {
+    let line = format!("{}\n", log_line.to_json().to_string());
+    match env::var("RMC_LOG_FILE") {
+        Err(_) => Ok(()),
+        Ok(path) => {
+            OpenOptions::new().append(true).create(true).open(path)?.write_all(line.as_bytes())
```

Comment:
What happens if this path is invalid or we don't have permission to write to it?

Reply:
I'm going to add a clear error message for this case which will also mention setting the `--save-logs` flag. If you think we should try to recover instead, I'm happy to hear your case!

Reply:
I am okay with a clear error message.

I would suggest to also check if we have permission to write to the file while parsing --save-logs so we can fail fast.